### PR TITLE
Correct the translation of protobuf to json at query frontend

### PR DESCRIPTION
**Pull request by annanay25**
Signed-off-by: Annanay <annanayagarwal@gmail.com>

**What this PR does**:
- This PR is a bugfix for #457, where we were incorrectly passing a json marshalled byte slice of traces to `combineTrace` which only accepts protobuf marshalled format.
- Changed signature of `combineTrace` to also return error, as it helps with bubbling it up as required
- Tests added accordingly

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Checklist**
- [x] Tests updated
- NA Documentation added
- NA `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: modules/compactor/compactor.go [hunk 1]
```diff
@@ -149,7 +149,8 @@ func (c *Compactor) Owns(hash string) bool {
 }
 
 func (c *Compactor) Combine(objA []byte, objB []byte) []byte {
-	return tempo_util.CombineTraces(objA, objB)
+	combinedTrace, _ := tempo_util.CombineTraces(objA, objB)
```
Comment (joe-elliott) [inline modules/compactor/compactor.go:152]: let's log this error

File: modules/frontend/frontend.go [hunk 2]
```diff
@@ -54,7 +55,14 @@ func NewTripperware(cfg Config, logger log.Logger, registerer prometheus.Registe
 			r = r.WithContext(ctx)
 			resp, err := rt.RoundTrip(r)
 
-			level.Info(logger).Log("method", r.Method, "url", r.URL.RequestURI(), "duration", time.Since(start).String(), "error", err == nil)
+			traceID, _ := middleware.ExtractTraceID(ctx)
+			var statusCode int
+			if resp == nil {
```
Comment (joe-elliott) [inline modules/frontend/frontend.go:60]: definitely a weird situation where resp is nil.  for clarity:

```
statusCode := 500
if resp != nil {
  statusCode = resp.StatusCode
}
```

File: modules/frontend/querysharding.go [hunk 3]
```diff
@@ -174,9 +196,31 @@ func mergeResponses(rrs []RequestResponse) (*http.Response, error) {
 	}
 
 	if errCode == http.StatusOK {
+		if marshallingFormat == util.ProtobufTypeHeaderValue {
```
Comment (joe-elliott) [inline modules/frontend/querysharding.go:199]: if there's a sane way to combine this with the querier code that also uses the Accept header to choose how to marshal, please do.  otherwise this is fine.

Comment (joe-elliott) [reply modules/frontend/querysharding.go:199]: Maybe could be cleaned up like:

```
var r Reader

if format == Proto {
 r := protoThing
} else {
  r := jsonThing
}

return &http.Response{  
  Body: Use r
 }
```

File: modules/ingester/instance.go [hunk 4]
```diff
@@ -337,7 +337,8 @@ func (i *instance) writeTraceToHeadBlock(id common.ID, b []byte) error {
 }
 
 func (i *instance) Combine(objA []byte, objB []byte) []byte {
-	return util.CombineTraces(objA, objB)
+	combinedTrace, _ := util.CombineTraces(objA, objB)
```
Comment (joe-elliott) [inline modules/ingester/instance.go:340]: log the error if it's set

File: pkg/util/trace_test.go [hunk 5]
```diff
@@ -71,7 +71,8 @@ func TestCombine(t *testing.T) {
 	}
```
Comment (joe-elliott) [inline pkg/util/trace_test.go:71]: add a test that fails and check that err != nil on return.

**Review by joe-elliott — CHANGES_REQUESTED** [review]
Couple of changes requested and then the linter is mad you didn't check err on some tests.

File: modules/compactor/compactor.go (hunk 1 above)
Comment (annanay25) [reply modules/compactor/compactor.go:152]: Hmm, as of now it will be logged at `CombineTrace` and in the parent fn, should we remove logging at `CombineTrace` altogether?

File: modules/ingester/instance.go (hunk 4 above)
Comment (annanay25) [reply modules/ingester/instance.go:340]: https://github.com/grafana/tempo/pull/480/files#r565891495

File: modules/frontend/querysharding.go (hunk 3 above)
Comment (annanay25) [reply modules/frontend/querysharding.go:199]: Have been thinking about a function like so - 

```
func MarshalTraceObject(trace *tempopb.Trace, marshallingFormat string) ([]byte, error)
```

But while this could be useful at the querier, at the frontend we receive the response in bytes, its not beneficial to unmarshal that into the object and then remarshal it at the function if we're going to return proto bytes anyway..

File: modules/compactor/compactor.go (hunk 1 above)
Comment (joe-elliott) [reply modules/compactor/compactor.go:152]: yeah, if we're going to return an error from CombineTraces(), let's log it outside that function.  honestly it's weird that a utility function like CombineTraces logs anything.

**Review by joe-elliott — APPROVED** [review]
One small request, but lgtm.
